File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/DayTwoViewModel.java
package droiddevelopers254.droidconke.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;

import droiddevelopers254.droidconke.datastates.SessionsState;
import droiddevelopers254.droidconke.repository.DayTwoRepo;
import droiddevelopers254.droidconke.repository.RoomStarrSessionRepo;

public class DayTwoViewModel extends ViewModel{
    private MediatorLiveData<SessionsState> sessionsStateMediatorLiveData;
    private DayTwoRepo dayTwoRepo;
    private RoomStarrSessionRepo roomStarrSessionRepo;

    public DayTwoViewModel(){
        dayTwoRepo = new DayTwoRepo();
        sessionsStateMediatorLiveData = new MediatorLiveData<>();
        roomStarrSessionRepo = new RoomStarrSessionRepo();
    }

    public LiveData<SessionsState> getSessions(){
        return sessionsStateMediatorLiveData;
    }

    public void getDayTwoSessions(){
        final LiveData<SessionsState> sessionsStateLiveData = dayTwoRepo.getDayTwoSessions();
        sessionsStateMediatorLiveData.addSource(sessionsStateLiveData,
                sessionsStateMediatorLiveData ->{
                    if (this.sessionsStateMediatorLiveData.hasActiveObservers()){
                        this.sessionsStateMediatorLiveData.removeSource(sessionsStateLiveData);
                    }
                    this.sessionsStateMediatorLiveData.setValue(sessionsStateMediatorLiveData);
                });
    }

}

File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/DayTwoViewModel.kt
package droiddevelopers254.droidconke.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import droiddevelopers254.droidconke.datastates.Result
import droiddevelopers254.droidconke.models.SessionsModel
import droiddevelopers254.droidconke.repository.DayTwoRepo
import kotlinx.coroutines.launch

class DayTwoViewModel(private val dayTwoRepo: DayTwoRepo) : ViewModel() {
  private val sessionsStateMediatorLiveData = MutableLiveData<List<SessionsModel>>()
  private val sessionError = MutableLiveData<String>()

  fun getSessionsResponse(): LiveData<List<SessionsModel>> = sessionsStateMediatorLiveData

  fun getSessionsError(): LiveData<String> = sessionError

  fun getDayTwoSessions() {
    viewModelScope.launch {
      when (val value = dayTwoRepo.getDayTwoSessions()) {
          is Result.Success -> sessionsStateMediatorLiveData.postValue(value.data)
          is Result.Error -> sessionError.postValue(value.exception)
      }
    }


  }
}
